task_3/counter.cpp

```diff
@@ -0,0 +1,324 @@
+#include "counter.h"
+
+#include <iostream>
+#include <boost/bind.hpp>
+#include <boost/asio.hpp>
+#include <boost/thread/thread.hpp>
+#include <experimental/filesystem>
+#include <chrono>
+#include <queue>
+#include <fstream>
+
+using std::string;
+using std::cout;
+using std::endl;
+using namespace std::experimental::filesystem;
+
+boost::mutex global_stream_lock;
+
+	Counter::Counter(path dir, string outFile) : _dir {dir}, _outFile {outFile}, finishedSearching {false}, bunchFilesToPush {2}
+		, bunchFilesToProcess {2} {	}
+
+	void Counter::searchFiles() {
+		global_stream_lock.lock();
+		std::cout << "[" << boost::this_thread::get_id() << "] searchFiles thread" << std::endl;
+		global_stream_lock.unlock();
+
+		std::queue<path> _files;
+		int n = 0;
+		// cycle analizing input directory and pushing decent files to queue
+		for(auto& p: std::experimental::filesystem::recursive_directory_iterator(_dir)) {
+			if (std::experimental::filesystem::is_regular_file(p) &&
+				(path(p).extension() == ".c" || path(p).extension() == ".cpp" ||
+				 path(p).extension() == ".h" || path(p).extension() == ".hpp") )
```

Comment:
This is hardcoded comparisons. Here better solution is to use a vector with extensions and any_of algorithm though in prospects the extensions can be easily added.

Reply:
agree with remark
I used std::set to hold all extentions, because easier way check presence any item in this container